### DISCUSSION TITLE
Remix vs Modern JS

### DISCUSSION THREAD

**Discussion by AsuraKev** [remix-run/remix#7767] [Q&A]
I was experimenting with Modern JS and it was looking good as it has first class support for MFE and upcoming integration with Module Federation and Rspack which is pretty exciting. Then I come across Remix which looks very similar but more adoption by people. 

Can someone explain which one to go for and why for the decision? :)

thanks

**Comment by sergiodxa** [comment]
What is ModernJS?

**Reply by AsuraKev** [reply]
https://modernjs.dev/en/

Next js like react framework, ability to have file system routing and built as SPA and SSR :)

**Comment by jrestall** [comment]
@sergiodxa modern.js (https://modernjs.dev/) is a web framework by the company behind TikTok.

@AsuraKev My high-level understanding of modern.js is that it is mostly a re-implementation of Remix on top of the webpack and Rspack compilers. It [copies](https://github.com/web-infra-dev/modern.js/blob/main/packages/cli/plugin-data-loader/src/cli/data.ts) and/or [references](https://github.com/web-infra-dev/modern.js/blob/13afea72dbb3a9fd0e49b537c58552fbb955bca1/packages/cli/plugin-data-loader/src/runtime/index.ts#L3) a lot of Remix code and packages to implement the same concepts, whilst [re-exporting react-router-dom](https://github.com/web-infra-dev/modern.js/blob/main/packages/toolkit/runtime-utils/src/router.ts) and [remix-router](https://github.com/web-infra-dev/modern.js/blob/13afea72dbb3a9fd0e49b537c58552fbb955bca1/packages/toolkit/runtime-utils/src/remixRouter.ts#L2) for the majority of its functionality. 

I'm not at all saying this is a bad thing as micro-frontends are obviously very important to them at their scale, so with Remix not supporting them, that's the path they had to go down along with giving them full control over their server, compiler and framework.

However if you are interested in Module Federation then I assume you are looking to build a very large website with very many teams. Personally I'd prefer the foundation of that website to be controlled by the original framework team, rather than a wrapper or re-implementation. For example when middleware or clientLoader functionality land in @remix-run/router, you can use it immediately rather than waiting for the modern.js team to support it.

It's all tradeoffs that you'll have to evaluate as Remix doesn't have anywhere near as seamless micro-frontend or module federation support, so if that's really important to you then maybe ModernJS is the best option. If you want to be on full blown Module Federation which is only WebPack and soon Rspack then it is probably the only option as I can't recommend Remix's [webpack compiler](https://github.com/remix-run/remix-webpack-demo/tree/main) as a good long-term option. Side note is that Vite announced a desire to support Module Federation in their new Rolldown bundler that will share code with Rspack, but that's probably a year away.

With Remix switching to Vite, you have some alternative options though. If you are happy with the more limited features and smaller community of the tooling agnostic and browser import maps based [Native Federation](https://www.angulararchitects.io/en/blog/announcing-native-federation-1-0/) and have the resources and skilled JS devs to manage that yourself then have a look at https://github.com/jrestall/remix-federation and ideally contribute to that so that Remix has a similarly good micro-frontend story. I don't see anything in ModernJS that isn't also achievable in Native Federation + Vite, they just have more helpers and orchestration around it to make it simple. There's also https://github.com/originjs/vite-plugin-federation but I preferred the browser native and agnostic approach.

**Comment by yimingjfe** [comment]
Hi, I am a member of Modern.js team. I want to share some ideas from my perspective.

Firstly I want to say that Remix is a fantastic framework. Modern.js was heavily inspired by Remix.js in routing and SSR. At the same time, Modern.js' routing solution is based on react-router, which is created by the Remix team.

There are some thoughts when we creating Modern.js:
1. We uses webpack and Rspack as bundlers in Modern.js. We have seen the perfect ecosystem of webpack and we hope that such an ecosystem can continue to develop. We're trying to make this ecosystem faster than ever with the help of Rspack.

2. In the enterprise-level application scenarios we face, micro-frontends are a very important way for us to solve the scaling problem, so we have made first-class support for micro-frontends in Modern.js. This also includes support for Module Federation.abs

3. We see that CSR (client-side rendering), SSR (server-side rendering) and SSG（Static Site Generation）each have their suitable usage scenarios. In some cases, using CSR mode can be more efficient and have lower operating costs. Therefore, in Modern.js, We provide support for all these rendering modes, and allow developers to choose the render mode they needed, which makes the entire framework more progressive.
